cmd/tree-main.go

```diff
@@ -0,0 +1,279 @@
+	treeNext      = "│"
+	treeLevel     = "  "
+)
+
+// Structured message depending on the type of console.
+type treeMessage struct {
+	Entry        string
+	IsDir        bool
+	BranchString string
+}
+
+// Colorized message for console printing.
+func (t treeMessage) String() string {
+	//fmt.Printf("%s\n", console.Colorize("Dir", url))
+	entryType := "File"
+	if t.IsDir {
+		entryType = "Dir"
+	}
+	return fmt.Sprintf("%s%s", t.BranchString, console.Colorize(entryType, t.Entry))
+}
+
+// JSON'ified message for scripting.
+// Does No-op. JSON requests are redirected to `ls -r --json`
+func (r treeMessage) JSON() string {
+	return ""
+}
+
+var treeFlags = []cli.Flag{
+	cli.BoolFlag{
+		Name:  "files, f",
+		Usage: "includes files in tree",
+	},
+	cli.IntFlag{
+		Name:  "depth, d",
+		Usage: "sets the depth threshold",
+		Value: 1000,
+	},
+}
+
+// trees files and folders.
+var treeCmd = cli.Command{
+	Name:   "tree",
+	Usage:  "list buckets and objects in a tree format",
+	Action: mainTree,
+	Before: setGlobalsFromContext,
+	Flags:  append(treeFlags, globalFlags...),
+	CustomHelpTemplate: `NAME:
+  {{.HelpName}} - {{.Usage}}
+
+USAGE:
+  {{.HelpName}} [FLAGS] TARGET [TARGET ...]
+
+FLAGS:
+  {{range .VisibleFlags}}{{.}}
+  {{end}}
+EXAMPLES:
+   1. List buckets on Amazon S3 cloud storage in a tree format.
+      $ {{.HelpName}} myS3
+
+   2. List all buckets on Amazon S3 cloud storage in a tree format.
+      $ {{.HelpName}} myS3/mybucket/
+
+   3. List all buckets on Amazon S3 cloud storage on Microsoft Windows in a tree format.
+      $ {{.HelpName}} myS3\mybucket\
+   
+   4. List all buckets including the objects on Amazon S3 cloud storage in a tree format.
+      $ {{.HelpName}} -f myS3/mybucket/
+   
+   5. Set the depth of the tree for listing.
+      $ {{.HelpName}} -d 2 myS3/mybucket/
+`,
+}
+
+// checkTreeSyntax - validate all the passed arguments
+func checkTreeSyntax(ctx *cli.Context) {
+	args := ctx.Args()
+
+	if ctx.Int("depth") <= 0 {
+		fatalIf(errInvalidArgument().Trace(args...), "depth should have a value greater than 0")
+	}
+
+	if (args.Present()) && len(args) == 0 {
+		args = []string{"."}
+		return
+	}
+
+	for _, url := range args {
+		if !isURLPrefixExists(url, false) {
+			fatalIf(probe.NewError(errors.New("See 'mc tree -h' for help")),
+				"'"+url+"' is not a valid 'alias[/bucket-name ...]'")
+		}
+	}
+}
+
+// doTree - list all entities inside a folder in a tree format.
+func doTree(url string, level int, leaf bool, dirClosed map[int]bool, depth int, includeFiles bool) error {
+
+	targetAlias, targetURL, _ := mustExpandAlias(url)
+	if !strings.HasSuffix(targetURL, "/") {
+		targetURL += "/"
+	}
+
+	clnt, err := newClientFromAlias(targetAlias, targetURL)
+	fatalIf(err.Trace(targetURL), "Unable to initialize target `"+targetURL+"`.")
+
+	prefixPath := clnt.GetURL().Path
+	separator := string(clnt.GetURL().Separator)
+	prefixPath = strings.TrimSuffix(prefixPath, prefixPath[strings.LastIndex(prefixPath, separator)+1:])
```

Comment:
What are you trying to do here? `filepath.Dir()` or `filepath.Base()`

Reply:
filepath.Dir() is what i was trying. It is doing `filepath.Clean()` additionally. so i preferred this.